### PR TITLE
[BEAM-4152] Kinds & URNs for Session Windows

### PR DESCRIPTION
**Pull request by seanhagen**
Added the `WindowKind` and coder things so that the already existing session
windowing would work.

Before, was getting the following panic when trying to use a session window (
with a gap time of 5 minutes ):

```
panic: Unexpected windowing strategy: SES[5m0s]
```

This PR still needs some work though -- now the error that I get is this:

```
panic: reflect.StructOf: StructOf does not allow unexported fields
```

After diving into the runtime package and doing some debugging, this seems to be coming from the encoder trying to encode a `time.Time` value and choking on the private `zone` struct. I'm at a loss there, as none of the structs in the pipeline I'm working on use `time.Time` ( or use a struct that uses `time.Time` somewhere )

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | --- | --- | --- | ---
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python3_Verify/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/) <br> [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/) | --- | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

**Comment by seanhagen** [comment]
R: @herohde

**Comment by seanhagen** [comment]
I don't know if the new panic is related to the session window or the code in my pipeline, but I'm going to keep diving into the code today and try to figure this out.

**Comment by lostluck** [comment]
I'll take a look! @herohde hasn't worked on the project in some time, but I welcome his insight if he finds the time.

**Comment by seanhagen** [comment]
Panic trace after code change:
```
panic: reflect.StructOf: StructOf does not allow unexported fields                                       

goroutine 1 [running]:                              
reflect.runtimeStructField(0xc00056c6a8, 0x4, 0xc00056c6ac, 0x4, 0x1208480, 0xef9a40, 0x0, 0x0, 0x0, 0xc00056d878, ...)                                                                                            
        /usr/local/go/src/reflect/type.go:2779 +0x1de                                                    
reflect.StructOf(0xc000685040, 0x3, 0x4, 0x0, 0x0)  
        /usr/local/go/src/reflect/type.go:2378 +0x1fcb                                                   
github.com/apache/beam/sdks/go/pkg/beam/core/runtime/graphx.decodeType(0xc00013ed00, 0x1b03c20, 0xc000099080, 0x1208480, 0xc000099080)                                                                             
        /home/sean/Code/DevOps/beam/sdks/go/pkg/beam/core/runtime/graphx/serialize.go:560 +0x954         
github.com/apache/beam/sdks/go/pkg/beam/core/runtime/graphx.decodeType(0xc00013ec80, 0x0, 0x0, 0x0, 0x1) 
        /home/sean/Code/DevOps/beam/sdks/go/pkg/beam/core/runtime/graphx/serialize.go:535 +0x41f         
github.com/apache/beam/sdks/go/pkg/beam/core/runtime/graphx.decodeType(0xc00013eb80, 0x7fb9d0a28000, 0x7fb9d0a28000, 0x0, 0x40d96f)                                                                                
        /home/sean/Code/DevOps/beam/sdks/go/pkg/beam/core/runtime/graphx/serialize.go:544 +0x625         
github.com/apache/beam/sdks/go/pkg/beam/core/runtime/graphx.decodeType(0xc00013eb00, 0xc000155490, 0x1, 0x1, 0x2)                                                                                                  
        /home/sean/Code/DevOps/beam/sdks/go/pkg/beam/core/runtime/graphx/serialize.go:585 +0xfc1         
github.com/apache/beam/sdks/go/pkg/beam/core/runtime/graphx.decodeType(0xc00013e980, 0x1, 0x2, 0xc00056d848, 0x1)                                                                                                  
        /home/sean/Code/DevOps/beam/sdks/go/pkg/beam/core/runtime/graphx/serialize.go:544 +0x625         
github.com/apache/beam/sdks/go/pkg/beam/core/runtime/graphx.decodeType(0xc00013e000, 0x198f, 0x198f, 0x198f, 0x0)                                                                                                  
        /home/sean/Code/DevOps/beam/sdks/go/pkg/beam/core/runtime/graphx/serialize.go:544 +0x625         
github.com/apache/beam/sdks/go/pkg/beam/core/runtime/graphx.decodeCustomCoder(0xc00082c7b0, 0x2214, 0x11f6e00, 0xc00082c7b0)                                                                                       
        /home/sean/Code/DevOps/beam/sdks/go/pkg/beam/core/runtime/graphx/serialize.go:146 +0x44          
github.com/apache/beam/sdks/go/pkg/beam/core/runtime/graphx.(*CoderUnmarshaller).makeCoder(0xc000506060, 0xc00056e040, 0x1, 0x1, 0x1)                                                                              
        /home/sean/Code/DevOps/beam/sdks/go/pkg/beam/core/runtime/graphx/coder.go:229 +0x170a            
github.com/apache/beam/sdks/go/pkg/beam/core/runtime/graphx.(*CoderUnmarshaller).Coder(0xc000506060, 0xc00056c098, 0x2, 0xc00056e040, 0x0, 0x0)                                                                    
        /home/sean/Code/DevOps/beam/sdks/go/pkg/beam/core/runtime/graphx/coder.go:115 +0x1c8             
github.com/apache/beam/sdks/go/pkg/beam/core/runtime/graphx.(*CoderUnmarshaller).makeCoder(0xc000506060, 0xc00056e0c0, 0x1, 0x1, 0x1)                                                                              
        /home/sean/Code/DevOps/beam/sdks/go/pkg/beam/core/runtime/graphx/coder.go:203 +0x68d             
github.com/apache/beam/sdks/go/pkg/beam/core/runtime/graphx.(*CoderUnmarshaller).Coder(0xc000506060, 0xc00056c120, 0x2, 0xc000494d50, 0xc000684b88, 0xc000684750)                                                  
        /home/sean/Code/DevOps/beam/sdks/go/pkg/beam/core/runtime/graphx/coder.go:115 +0x1c8             
github.com/apache/beam/sdks/go/pkg/beam/runners/dataflow/dataflowlib.(*translator).translateCoder(0xc000506080, 0xc0001e64e0, 0xc00056c120, 0x2, 0xc0006847e8)                                                     
        /home/sean/Code/DevOps/beam/sdks/go/pkg/beam/runners/dataflow/dataflowlib/translate.go:310 +0x47 
github.com/apache/beam/sdks/go/pkg/beam/runners/dataflow/dataflowlib.(*translator).translateOutputs(0xc000506080, 0xc000494d50, 0x11043b7, 0xb, 0xc00056c141)                                                      
        /home/sean/Code/DevOps/beam/sdks/go/pkg/beam/runners/dataflow/dataflowlib/translate.go:292 +0x190                                                                                                          
github.com/apache/beam/sdks/go/pkg/beam/runners/dataflow/dataflowlib.(*translator).translateTransform(0xc000506080, 0x0, 0x0, 0xc0005e9a28, 0x2, 0xc000506080, 0xc0004d7828, 0xe006c6, 0x10353a0, 0xc000506080)    
        /home/sean/Code/DevOps/beam/sdks/go/pkg/beam/runners/dataflow/dataflowlib/translate.go:110 +0x10f                                                                                                          
github.com/apache/beam/sdks/go/pkg/beam/runners/dataflow/dataflowlib.(*translator).translateTransforms(0xc000506080, 0x0, 0x0, 0xc00044efa0, 0x5, 0x5, 0xc0004d78b0, 0x42c1a1, 0x113f9e8, 0xc0004d78c0, ...)       
        /home/sean/Code/DevOps/beam/sdks/go/pkg/beam/runners/dataflow/dataflowlib/translate.go:96 +0xd1  
github.com/apache/beam/sdks/go/pkg/beam/runners/dataflow/dataflowlib.translate(0xc00044ee60, 0xc0004d78c0, 0xc0004d7960, 0x4f430d, 0xc0000c2000, 0xc0006fa000)                                                     
        /home/sean/Code/DevOps/beam/sdks/go/pkg/beam/runners/dataflow/dataflowlib/translate.go:72 +0x76  
github.com/apache/beam/sdks/go/pkg/beam/runners/dataflow/dataflowlib.Translate(0xc00044ee60, 0xc0001340c0, 0xc00049c300, 0x31, 0xc0005f7710, 0x30, 0x1, 0xad47, 0xad47)                                            
        /home/sean/Code/DevOps/beam/sdks/go/pkg/beam/runners/dataflow/dataflowlib/job.go:66 +0x43        
github.com/apache/beam/sdks/go/pkg/beam/runners/dataflow/dataflowlib.Execute(0x11ff040, 0xc0000401c8, 0xc0006234a0, 0xc0001340c0, 0xc00049c300, 0x31, 0xc0005f7710, 0x30, 0x0, 0x0, ...)                           
        /home/sean/Code/DevOps/beam/sdks/go/pkg/beam/runners/dataflow/dataflowlib/execute.go:79 +0x58d   
github.com/apache/beam/sdks/go/pkg/beam/runners/dataflow.Execute(0x11ff040, 0xc0000401c8, 0xc00000e6e0, 0x8, 0xc00047a8a8)                                                                                         
        /home/sean/Code/DevOps/beam/sdks/go/pkg/beam/runners/dataflow/dataflow.go:159 +0x9c5             
github.com/apache/beam/sdks/go/pkg/beam.Run(0x11ff040, 0xc0000401c8, 0x7fffb78c2a78, 0x8, 0xc00000e6e0, 0xc0004aa540, 0xc0004ad0e0)                                                                                
        /home/sean/Code/DevOps/beam/sdks/go/pkg/beam/runner.go:50 +0x8e                                  
github.com/apache/beam/sdks/go/pkg/beam/x/beamx.Run(0x11ff040, 0xc0000401c8, 0xc00000e6e0, 0xc0000c2f00, 0x111b109)                                                                                                
        /home/sean/Code/DevOps/beam/sdks/go/pkg/beam/x/beamx/run.go:42 +0x57 
```

**Comment by lostluck** [comment]
OK! So the panic is because the type isn't registered with beam.RegisterType(), which is an error message that will be improving with the following PR https://github.com/apache/beam/pull/8064

The next session windowing error you'll run into will be on the "exec" side that happens on the workers. It's unaware of Session windows. You'll note this is where the harder bit of assigning windows happens.

https://github.com/apache/beam/blob/master/sdks/go/pkg/beam/core/runtime/exec/window.go#L57

I found the outstanding JIRA issue for session windowing, https://issues.apache.org/jira/browse/BEAM-4152  
Merging windows is not handled on the SDK side yet either (otherwise this would be done already), but the answer here is "see what Java and/python are doing, and do that in the Go way".

**Comment by lostluck** [comment]
@lostluck  Ating myself so I can see this in my mentions, and not lose it.

**Comment by lostluck** [comment]
I've conferred with someone who knows the Java side better than me, and the exec stack side should be pretty straight forward, because sessions are known. You shouldn't need to implement any merge logic on the SDK side.

https://github.com/apache/beam/blob/master/sdks/go/pkg/beam/core/runtime/exec/window.go#L57 should be very similar to fixed window. 

The other next spot would be to ensure that there's a coder for the windows from the exec side:
You can see how the others are implemented here:
https://github.com/apache/beam/blob/master/sdks/go/pkg/beam/core/runtime/exec/coder.go#L274 

The Runner in principle should handle the merging following a GroupByKey/Combine operation.

Note: This wouldn't be handled properly in the Go SDK direct runner, that's where the sorting and merging would need to happen based on the configured gap size.

**Comment by lostluck** [comment]
Run Go Postcommit

**Comment by ptomasroos** [comment]
@lostluck what is the reason for having the direct runner now that the universal runner will exist? 
Is the idea to remove the language specific direct runners in the end when the universal runner is good enough?

**Comment by robertwb** [comment]
Direct runners still have value in being able to run entirely in a single
process, in the user's preferred language/environment (for both development
and testing). Exactly how complete to make a language-specific direct
runner depends on how much value people find in it (and how much negative
value people place on taking on a cross-language dependency).

**Comment by ptomasroos** [comment]
I definitely agree with that @robertwb, it was more a question if having a direct runner becomes a constraint for evolving the go sdk. Since the translation / graph building phase is a lot different than the execution part of a direct runner. But point taken. Thanks

**Comment by lostluck** [comment]
@ptomasroos At present the "testing and experimenting" path that works for the Go SDK is based on the direct runner, as @robertwb mentions. 

Right now the direct runner is a constraint since work hasn't been done to use the Go SDK with the (new?) Python ULR, though I do know it was working with the Java ULR at some point, let alone a free runner such as Flink. Otherwise, the only thing that is tested via the post commits is Dataflow, but that's not great for users experimenting, at the end of the day since it costs money, which isn't ideal for users just wanting to try things.

https://issues.apache.org/jira/browse/BEAM-6958 is a task to split out the integration tests in the post commit rubric. Which would be natural starting point for demonstrating how to use the SDK against other runners ensuring that the SDK works on said runners to boot. That should help a touch, once the jenkin's machine apocalypse is over.

**Comment by robertwb** [comment]
What remains to run Go tests against the Python ULR? It should just be a matter of starting up the ULR and pointing to it, right?

**Comment by lostluck** [comment]
@robertwb That's my understanding. Of course also the time required to do so as well.

**Comment by robertwb** [comment]
Yes, that should be sufficient.

File: sdks/go/pkg/beam/core/runtime/graphx/coder.go [hunk 1]
```diff
@@ -39,6 +39,7 @@ const (
 
 	urnGlobalWindow   = "beam:coder:global_window:v1"
 	urnIntervalWindow = "beam:coder:interval_window:v1"
+	urnSessionWindow  = "beam:coder:session_window:v1"
```
Comment (lukecwik) [inline sdks/go/pkg/beam/core/runtime/graphx/coder.go]: If you want to use a well known name which isn't scoped to Go, you'll want to add it to the set here https://github.com/apache/beam/blob/e0cbd2aa7e371c75511544ab78075d54f3f086ca/model/pipeline/src/main/proto/beam_runner_api.proto#L616

The other well known window coders don't do a good job of it but you should really specify the encoding, look at beam:coder:iterable:v1 or beam:coder:timer:v1 for examples.

**Pull request closed without merging**